File: flutter/lib/services/agent_event_parser.dart
import 'dart:convert';

import 'package:agora_rtm/agora_rtm.dart';

class AgentStateChange {
  const AgentStateChange({
    required this.agentUserId,
    required this.state,
    required this.turnId,
    required this.timestamp,
  });

  final String agentUserId;
  final String state;
  final int turnId;
  final int timestamp;
}

class AgentError {
  const AgentError({
    required this.agentUserId,
    required this.module,
    required this.code,
    required this.message,
    required this.timestamp,
    this.turnId,
  });

  final String agentUserId;
  final String module;
  final int code;
  final String message;
  final int timestamp;
  final int? turnId;
}

class AgentEventParser {
  static String formatConsoleMessage(String rawMessage) {
    return rawMessage.replaceAllMapped(
      RegExp(r'\\u([0-9a-fA-F]{4})'),
      (match) {
        final int? codeUnit = int.tryParse(match.group(1)!, radix: 16);
        if (codeUnit == null) {
          return match.group(0)!;
        }
        return String.fromCharCode(codeUnit);
      },
    );
  }

  static AgentStateChange? parsePresenceEvent(
    PresenceEvent event, {
    required String currentChannelName,
    required int lastTurnId,
    required int lastTimestamp,
  }) {
    if (event.channelName != currentChannelName) {
      return null;
    }
    if (event.channelType != RtmChannelType.message) {
      return null;
    }
    if (event.type != RtmPresenceEventType.remoteStateChanged) {
      return null;
    }

    final Map<String, String> stateMap = <String, String>{
      for (final StateItem item in event.stateItems ?? <StateItem>[])
        if (item.key != null && item.value != null) item.key!: item.value!,
    };

    final String state = (stateMap['state'] ?? '').trim().toLowerCase();
    if (state.isEmpty) {
      return null;
    }

    final int turnId = int.tryParse(stateMap['turn_id'] ?? '') ?? 0;
    final int timestamp = event.timestamp ?? 0;
    if (turnId < lastTurnId) {
      return null;
    }
    if (timestamp <= lastTimestamp) {
      return null;
    }

    return AgentStateChange(
      agentUserId: event.publisher ?? '',
      state: state,
      turnId: turnId,
      timestamp: timestamp,
    );
  }

  static AgentError? parseMessageError(
    String rawMessage, {
    required String agentUserId,
  }) {
    try {
      final dynamic decoded = jsonDecode(rawMessage);
      if (decoded is! Map<String, dynamic>) {
        return null;
      }
      if ((decoded['object'] ?? '').toString() != 'message.error') {
        return null;
      }

      return AgentError(
        agentUserId: agentUserId,
        module: (decoded['module'] ?? '').toString(),
        code: _asInt(decoded['code']) ?? -1,
        message: (decoded['message'] ?? 'Unknown error').toString(),
        timestamp: _asInt(decoded['send_ts']) ?? 0,
        turnId: _asInt(decoded['turn_id']),
      );
    } catch (_) {
      return null;
    }
  }

  static int? _asInt(dynamic value) {
    if (value == null) {
      return null;
    }
    if (value is int) {
      return value;
    }
    if (value is num) {
      return value.toInt();
    }
    if (value is String) {
      return int.tryParse(value);
    }
    return null;
  }
}

File: unity/Assets/Scripts/Quickstart/AgentEventParser.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;
using Agora.Rtm;
using UnityEngine;

namespace Quickstart
{
    [Serializable]
    public class AgentStateChange
    {
        public string AgentUserId;
        public string State;
        public int TurnId;
    }

    [Serializable]
    public class AgentError
    {
        public string AgentUserId;
        public string Module;
        public int Code;
        public string Message;
        public long Timestamp;
        public int? TurnId;
    }

    [Serializable]
    public class AgentInterrupt
    {
        public string AgentUserId;
        public int TurnId;
    }

    public static class AgentEventParser
    {
        private static readonly HashSet<string> SupportedPresenceStates = new HashSet<string>
        {
            "silent",
            "listening",
            "thinking",
            "speaking",
        };

        public static string FormatConsoleMessage(string rawMessage)
        {
            return Regex.Replace(
                rawMessage,
                "\\\\u([0-9a-fA-F]{4})",
                match =>
                {
                    if (!int.TryParse(match.Groups[1].Value, System.Globalization.NumberStyles.HexNumber, null, out var codeUnit))
                    {
                        return match.Value;
                    }
                    return ((char)codeUnit).ToString();
                }
            );
        }

        public static string ParseMessageType(string rawMessage)
        {
            try
            {
                var payload = JsonUtility.FromJson<MessageEnvelope>(rawMessage);
                if (payload == null) return string.Empty;
                return !string.IsNullOrEmpty(payload.@object)
                    ? payload.@object
                    : payload.event_type ?? string.Empty;
            }
            catch
            {
                return string.Empty;
            }
        }

        public static string FormatPresenceEvent(PresenceEvent @event)
        {
            if (@event == null) return "PresenceEvent(null)";

            var builder = new System.Text.StringBuilder();
            builder.Append("PresenceEvent { ");
            builder.Append("type=").Append(@event.type);
            builder.Append(", channelType=").Append(@event.channelType);
            builder.Append(", channelName=").Append(@event.channelName ?? string.Empty);
            builder.Append(", publisher=").Append(@event.publisher ?? string.Empty);

            builder.Append(", stateItems=[");
            if (@event.stateItems != null)
            {
                for (var i = 0; i < @event.stateItems.Length; i++)
                {
                    var item = @event.stateItems[i];
                    if (item == null) continue;
                    if (i > 0) builder.Append(", ");
                    builder.Append(item.key ?? string.Empty).Append("=").Append(item.value ?? string.Empty);
                }
            }
            builder.Append("]");

            if (@event.interval != null)
            {
                builder.Append(", interval=");
                builder.Append("{ joinCount=").Append(@event.interval.joinUserList?.Length ?? 0);
                builder.Append(", leaveCount=").Append(@event.interval.leaveUserList?.Length ?? 0);
                builder.Append(", timeoutCount=").Append(@event.interval.timeoutUserList?.Length ?? 0);
                builder.Append(", userStateCount=").Append(@event.interval.userStateList?.Length ?? 0);
                builder.Append(" }");
            }

            if (@event.snapshot != null)
            {
                builder.Append(", snapshot=");
                builder.Append("{ userStateCount=").Append(@event.snapshot.userStateList?.Length ?? 0);
                builder.Append(" }");
            }

            builder.Append(" }");
            return builder.ToString();
        }

        public static AgentStateChange ParsePresenceEvent(
            PresenceEvent @event,
            string currentChannelName,
            int lastTurnId,
            string lastState
        )
        {
            if (@event == null) return null;
            if (@event.channelName != currentChannelName) return null;
            if (@event.channelType != RTM_CHANNEL_TYPE.MESSAGE) return null;
            if (@event.type != RTM_PRESENCE_EVENT_TYPE.REMOTE_STATE_CHANGED) return null;

            var stateMap = new Dictionary<string, string>();
            if (@event.stateItems != null)
            {
                foreach (var item in @event.stateItems)
                {
                    if (item == null || string.IsNullOrEmpty(item.key)) continue;
                    stateMap[item.key] = item.value ?? string.Empty;
                }
            }

            var state = (stateMap.TryGetValue("state", out var stateValue) ? stateValue : string.Empty).Trim().ToLowerInvariant();
            if (string.IsNullOrEmpty(state)) return null;
            if (!SupportedPresenceStates.Contains(state)) return null;

            var turnId = 0;
            if (stateMap.TryGetValue("turn_id", out var turnIdValue))
            {
                int.TryParse(turnIdValue, out turnId);
            }

            if (turnId < lastTurnId) return null;
            if (turnId == lastTurnId && state == (lastState ?? string.Empty).Trim().ToLowerInvariant()) return null;

            return new AgentStateChange
            {
                AgentUserId = @event.publisher ?? string.Empty,
                State = state,
                TurnId = turnId,
            };
        }

        public static AgentInterrupt ParseMessageInterrupt(string rawMessage, string agentUserId)
        {
            try
            {
                var payload = JsonUtility.FromJson<MessageInterruptPayload>(rawMessage);
                if (payload == null) return null;
                if ((payload.@object ?? string.Empty) != "message.interrupt") return null;

                return new AgentInterrupt
                {
                    AgentUserId = agentUserId ?? string.Empty,
                    TurnId = payload.turn_id,
                };
            }
            catch
            {
                return null;
            }
        }

        public static AgentError ParseMessageError(string rawMessage, string agentUserId)
        {
            try
            {
                var payload = JsonUtility.FromJson<MessageErrorPayload>(rawMessage);
                if (payload == null) return null;
                if ((payload.@object ?? string.Empty) != "message.error") return null;

                return new AgentError
                {
                    AgentUserId = agentUserId ?? string.Empty,
                    Module = payload.module ?? string.Empty,
                    Code = payload.code,
                    Message = string.IsNullOrEmpty(payload.message) ? "Unknown error" : payload.message,
                    Timestamp = payload.send_ts,
                    TurnId = payload.turn_id > 0 ? payload.turn_id : (int?)null,
                };
            }
            catch
            {
                return null;
            }
        }

        [Serializable]
        private class MessageErrorPayload
        {
            public string @object;
            public string module;
            public int code;
            public string message;
            public long send_ts;
            public int turn_id;
        }

        [Serializable]
        private class MessageInterruptPayload
        {
            public string @object;
            public int turn_id;
        }

        [Serializable]
        private class MessageEnvelope
        {
            public string @object;
            public string event_type;
        }
    }
}
